include/deal.II/hp/refinement.h

```diff
@@ -158,40 +158,85 @@ namespace hp
                             const std::vector<bool> &            p_flags);
 
     /**
-     * Adapt the finite element on cells whose smoothness indicators meet a
-     * certain threshold.
-     *
-     * The threshold will be chosen for refined and coarsened cells
-     * individually. For each class of cells, we determine the maximal and
-     * minimal values of the smoothness indicators and determine the threshold
-     * by linear interpolation between these limits. Parameters
-     * @p p_refine_fraction and @p p_refine_coarsen are used as interpolation
-     * factors, where `0` corresponds to the minimal and `1` to the maximal
-     * value. By default, mean values are considered as thresholds.
-     *
-     * We consider a cell for p-refinement if it is flagged for refinement and
-     * its smoothness indicator is larger than the corresponding threshold. The
-     * same applies for p-coarsening, but the cell's indicator must be lower
-     * than the threshold.
-     *
-     * Each entry of the parameter @p smoothness_indicators needs to correspond
-     * to an active cell. Parameters @p p_refine_fraction and
-     * @p p_coarsen_fraction need to be in the interval $[0,1]$.
+     * Adapt which finite element to use on cells whose criteria meet a certain
+     * absolute threshold.
+     *
+     * For p-refinement and p-coarsening, two separate thresholds need to
+     * provided via parameters @p p_refine_threshold and @p p_coarsen_threshold.
+     *
+     * We consider a cell for p-adaptivity if it is currently flagged for
+     * refinement or coarsening and its criterion successfully compares to the
+     * corresponding threshold. Let us be more specific on the default case: We
+     * consider a cell for p-refinement if it is flagged for refinement and its
+     * criterion is larger than the corresponding threshold. The same applies
+     * for p-coarsening, but the cell's criterion must be lower than the
+     * threshold. However, different compare function objects can be supplied
+     * via the parameters @p compare_refine and @p compare_coarsen to impose
+     * different decision strategies.
+     *
+     * Each entry of the parameter @p criteria needs to correspond to an active
+     * cell.
      *
      * @note Preceeding calls of Triangulation::prepare_for_coarsening_and_refinement()
      *   may change refine and coarsen flags, which will ultimately change the
      *   results of this function.
      */
     template <int dim, typename Number, int spacedim>
     void
-    p_adaptivity_from_threshold(
+    p_adaptivity_from_absolute_threshold(
       const hp::DoFHandler<dim, spacedim> &dof_handler,
-      const Vector<Number> &               smoothness_indicators,
+      const Vector<Number> &               criteria,
+      const Number                         p_refine_threshold,
+      const Number                         p_coarsen_threshold,
+      const std::function<bool(const Number &, const Number &)>
```

Comment:
So this would become as follows:
```suggestion
      const std::function<bool(const typename identity<Number>::type &, const typename identity<Number>::type &)>
```

Reply:
Looks a bit different now, but this is essentially what I did in the last commit with an alias :)